### PR TITLE
Added exporting cpu/mem request metrics and computing node_reservation metric based on them

### PR DESCRIPTION
**Pull request by piosz**

### CODE REVIEW

**Comment by piosz** [comment]
cc @vishh

**Comment by vishh** [comment]
add to whitelist

**Comment by mvdan** [comment]
LGTM

File: sources/api/types.go [hunk 1]
```diff
@@ -57,19 +57,27 @@ func (a *AggregateData) Merge(b *AggregateData) {
...
+type ContainerStats struct {
+	cadvisor.ContainerStats
+}
+
 type Container struct {
 	Hostname   string
 	ExternalID string
 	Name       string
 	Image      string
-	// TODO(vishh): Consider defining an internal Spec and Stats API to guard against
```
Comment (mwielgus) [inline sources/api/types.go]: This TODO should not be removed.

File: sources/api/types.go [hunk 2]
```diff
@@ -57,19 +57,27 @@ func (a *AggregateData) Merge(b *AggregateData) {
...
 }
 
+type ContainerSpec struct {
+	cadvisor.ContainerSpec
+	CpuRequest    int64
+	MemoryRequest int64
+}
+
+type ContainerStats struct {
+	cadvisor.ContainerStats
```
Comment (mwielgus) [inline sources/api/types.go]: What is the purpose of this struct?

File: sources/api/types.go (hunk 1 above)
Comment (vishh) [inline sources/api/types.go]: +1. We can get rid of cadvisor's Spec and Stats completely. The model does that.

File: sources/api/types.go [hunk 3]
```diff
@@ -57,19 +57,27 @@ func (a *AggregateData) Merge(b *AggregateData) {
 	a.Events = append(a.Events, b.Events...)
 }
 
+type ContainerSpec struct {
+	cadvisor.ContainerSpec
+	CpuRequest    int64
```
Comment (mwielgus) [inline sources/api/types.go]: Can we have these 2 grouped in a struct?
Would it be good to have CpuLimit as well?

File: sources/pods.go [hunk 4]
```diff
@@ -83,6 +83,12 @@ func (self *realPodsApi) parsePod(podNodePair *podNodePair) *api.Pod {
 		localContainer := api.Container{}
 		localContainer.Name = container.Name
 		localContainer.Image = container.Image
+		if cpu, ok := container.Resources.Requests[kapi.ResourceCPU]; ok {
+			localContainer.Spec.CpuRequest = cpu.MilliValue()
```
Comment (vishh) [inline sources/pods.go]: We will eventually need both request and limits I guess. Why not follow kube api semantics and make requests and limits a map[resourceType]Quantity?

File: sinks/api/supported_metrics.go [hunk 5]
```diff
@@ -109,26 +124,41 @@ var statMetrics = []SupportedStatMetric{
...
-		GetValue: func(spec *cadvisor.ContainerSpec, stat *cadvisor.ContainerStats) []InternalPoint {
+		GetValue: func(spec *source_api.ContainerSpec, stat *source_api.ContainerStats) []InternalPoint {
 			return []InternalPoint{{Value: int64(spec.Memory.Limit)}}
 		},
 		OnlyExportIfChanged: true,
 	},
 	{
 		MetricDescriptor: MetricDescriptor{
+			Name:        "memory/request",
+			Description: "Memory request",
```
Comment (vishh) [inline sinks/api/supported_metrics.go]: Should we clarify request vs limit? If we were to run with cadvisor in standalone mode, what will this metric be?

File: sources/api/types.go (hunk 3 above)
Comment (vishh) [inline sources/api/types.go]: Clarify what request means since it is currently specific to just kubernetes.

Comment (mwielgus) [inline sources/api/types.go]: Actually I've found cpu limit inside of cadvisor cpu spec. Adding the new cpu-related fields will create a mess. I would opt for removing cadvisor dependency in sources api (and completing the below TODO).

Comment (vishh) [inline sources/api/types.go]: That makes sense. We should handle both request and limit though. Values from cAdvisor will translate to limits, but that can be overridden in the case of Kubernetes.

File: sources/pods.go (hunk 4 above)
Comment (mwielgus) [inline sources/pods.go]: Limits are inside of cadvisor structs :/

File: sources/api/types.go (hunk 1 above)
Comment (piosz) [inline sources/api/types.go]: This is a good idea and a good candidate for 1.2 cleanup.

File: sources/api/types.go (hunk 2 above)
Comment (piosz) [inline sources/api/types.go]: To make it consistent with ContainerSpec above.

File: sources/api/types.go (hunk 3 above)
Comment (piosz) [inline sources/api/types.go]: Having these 2 grouped in a struct makes sense but probably the better option is what @vishh described in https://github.com/kubernetes/heapster/pull/539#discussion_r39319547.

As @mwielgus noticed cpu limit is already computed by cadvisor. Let's remove cadvisor dep in a new PR (probably after 1.1). AFAIK there is no requests computed by cadvisor so we need to workaround it by obtaining the data from Kubernetes.

File: sources/pods.go (hunk 4 above)
Comment (piosz) [inline sources/pods.go]: Good idea.

File: sinks/api/supported_metrics.go (hunk 5 above)
Comment (piosz) [inline sinks/api/supported_metrics.go]: Extended description. In the standalone mode no values will be exported here.

**Comment by piosz** [comment]
PTAL @mwielgus @vishh 

How about defining an internal Spec&Stats API in a follow up PR?

**Comment by mwielgus** [comment]
Please squash the commits.

**Comment by piosz** [comment]
I think the commits are meaningful so let's leave them.

**Comment by mwielgus** [comment]
LGTM

**Comment by vishh** [comment]
@piosz: SGTM.

File: sources/pods.go (hunk 4 above)
Comment (vishh) [inline sources/pods.go]: Introducing internal abstractions for stats and spec should solve this issue.

**Comment by vishh** [comment]
LGTM. Thanks @piosz

**Pull request merged**